src/async_runtime.rs

```diff
@@ -29,44 +30,83 @@ impl AsyncRuntime {
         Ok(Self { rt })
     }
 
-    pub fn handle(&self) -> Handle {
+    pub(crate) fn handle(&self) -> Handle {
         let handle = self.rt.handle().clone();
         Handle { handle }
     }
 }
 
 #[allow(dead_code)]
 impl Handle {
-    pub fn spawn<F>(&self, future: F) -> JoinHandle<F::Output>
+    pub(crate) fn spawn<F>(&self, future: F) -> JoinHandle<F::Output>
     where
         F: Future + Send + 'static,
         F::Output: Send + 'static,
     {
         self.handle.spawn(future)
     }
 
-    pub fn spawn_repeating_task<Func, F>(&self, interval: Duration, func: Func) -> JoinHandle<()>
+    pub(crate) fn spawn_repeating_task<Func, F>(
+        &self,
+        interval: Duration,
+        func: Func,
+    ) -> RepeatingTaskHandle
     where
         Func: Fn() -> F + Send + Sync + 'static,
         F: Future + Send + 'static,
         F::Output: Send + 'static,
     {
-        self.handle.spawn(async move {
+        let (stop_sender, mut stop_receiver) = mpsc::channel(1);
+        let (status_sender, status_receiver) = mpsc::channel(1);
+        let handle = self.handle.spawn(async move {
             let mut interval = time::interval(interval);
             interval.set_missed_tick_behavior(time::MissedTickBehavior::Skip);
             interval.tick().await;
             loop {
                 func().await;
-                interval.tick().await;
+                tokio::select! {
+                    _ = stop_receiver.recv() => {
+                        break;
+                    },
+                    _ = interval.tick() => {},
+                }
             }
-        })
+            drop(status_sender);
+        });
+        RepeatingTaskHandle {
+            handle,
+            stop_sender,
+            status_receiver,
+        }
     }
 
-    pub fn block_on<F: Future>(&self, future: F) -> F::Output {
+    pub(crate) fn block_on<F: Future>(&self, future: F) -> F::Output {
         self.handle.block_on(future)
     }
 }
 
+pub(crate) struct RepeatingTaskHandle {
+    handle: JoinHandle<()>,
+    stop_sender: mpsc::Sender<()>,
+    status_receiver: mpsc::Receiver<()>,
+}
+
+impl RepeatingTaskHandle {
+    pub(crate) fn safe_abort(&self) {
+        self.stop_sender.blocking_send(()).unwrap();
+    }
+
+    pub(crate) fn block_until_finished(&mut self) {
+        self.status_receiver.blocking_recv();
+    }
```

Comment:
What's the reason between separating `safe_abort()` and `block_until_finished`? Is there any use case for that, or would it make more sense to make `safe_abort()` blocking until finished itself? Or maybe add a third function that calls both of these functions and is called `safe_abort_blocking()` or something?

As a programmer using these functions I might think `safe_abort()` is itself already blocking and forget to `block_until_finished()`.

Reply:
I see your point. Right now, a blocking version would be enough, but I think there might be use cases in which we don't care in which order two or more tasks are stopped. In that case, it's more efficient to call`safe_abort()` on them and only wait after.

> As a programmer using these functions I might think safe_abort() is itself already blocking and forget to block_until_finished().

I think of `safe_abort()` as a variation of the original `JoinHandle<>.abort()`, and that is my main argument for not making it blocking.

I added the `safe_abort_blocking()` method in bb46507.